apps/vscode/src/main.ts

```diff
@@ -125,9 +128,46 @@ export async function activate(context: vscode.ExtensionContext) {
 
   // activate providers common to browser/node
   activateCommon(context, host, engine, commands);
+
+  // if positron
+  setEditorOpener();
+
+  vscode.workspace.onDidChangeConfiguration(async (event) => {
+    if (event.affectsConfiguration('quarto.defaultEditor')) {
+      setEditorOpener();
+    }
+  });
+
+  const documentOpenHandler = vscode.workspace.onDidOpenTextDocument(async (document: vscode.TextDocument) => {
+    // Check if the document language is "quarto"
+    const config = vscode.workspace.getConfiguration('quarto').get<string>('defaultEditor');
+    if (document.languageId != 'quarto') {
+      return;
+    }
+    if (suppressOpenHandler) {
+      suppressOpenHandler = false; // Reset the flag
+      return;
+    }
+    const editorMode = await determineMode(document.getText());
+    if (editorMode && editorMode != config) {
+      suppressOpenHandler = true;
+      const editorOpener = editorMode === 'visual' ? VisualEditorProvider.viewType : 'textEditor';
+      await vscode.commands.executeCommand('workbench.action.closeActiveEditor');
+      await vscode.commands.executeCommand("vscode.openWith",
+        document.uri,
+        editorOpener
+      );
+      return;
```

Comment:
There seems to be some kind of timing issue here - Even though we are in `onDidOpenTextDocument()`, it is often not the _active_ editor yet, so often it won't actually close the editor, and then `openWith` is a no-op because it thinks that URI is already open.